File: scripts/migrate-events-to-db.js
#!/usr/bin/env node

/**
 * Script to migrate events from JSON files to the database.
 * This script reads events from public/data/all_events_combined.json and inserts them into the database.
 */

const fs = require('fs').promises;
const path = require('path');
const { PrismaClient } = require('@prisma/client');

const prisma = new PrismaClient();

class EventMigrator {
  constructor() {
    this.dataFile = path.join(__dirname, '..', 'public', 'data', 'all_events_combined.json');
  }

  async loadEventsFromJson() {
    try {
      const fileContent = await fs.readFile(this.dataFile, 'utf8');
      const data = JSON.parse(fileContent);
      return data.events || [];
    } catch (error) {
      console.error(`Error reading ${this.dataFile}:`, error.message);
      return [];
    }
  }

  async createOrGetCommunity(communityName) {
    if (!communityName) return null;

    // Try to find existing community
    let existing = await prisma.community.findFirst({
      where: { name: communityName }
    });

    if (existing) {
      return existing.id;
    }

    // Create new community
    const newCommunity = await prisma.community.create({
      data: {
        name: communityName,
        description: `Community for ${communityName}`
      }
    });

    return newCommunity.id;
  }

  async createOrGetLocation(locationData) {
    if (!locationData || !locationData.address) return null;

    const address = locationData.address;

    // Try to find existing location
    let existing = await prisma.location.findFirst({
      where: { address: address }
    });

    if (existing) {
      return existing.id;
    }

    // Create new location
    const newLocation = await prisma.location.create({
      data: {
        name: locationData.name || 'Unknown Location',
        address: address,
        city: locationData.city || '',
        state: locationData.state || '',
        zip: locationData.zip || ''
      }
    });

    return newLocation.id;
  }

  async migrateEvent(eventData) {
    try {
      // Check if event already exists by externalId or name + startDate
      let existingEvent = null;
      
      if (eventData.externalId) {
        existingEvent = await prisma.event.findUnique({
          where: { externalId: eventData.externalId }
        });
      }

      if (!existingEvent) {
        // Try to find by name and start date
        const startDate = new Date(eventData.startDate);
        if (isNaN(startDate.getTime())) {
          console.log(`Skipping event with invalid date: ${eventData.name}`);
          return false;
        }
        
        existingEvent = await prisma.event.findFirst({
          where: {
            name: eventData.name,
            startDate: startDate
          }
        });
      }

      if (existingEvent) {
        console.log(`Event already exists: ${eventData.name}`);
        return true;
      }

      // Create or get community
      let communityId = null;
      if (eventData.community) {
        communityId = await this.createOrGetCommunity(eventData.community);
      }

      // Create or get location
      let locationId = null;
      if (eventData.location) {
        locationId = await this.createOrGetLocation(eventData.location);
      }

      // Parse dates
      const startDate = new Date(eventData.startDate);
      const endDate = eventData.endDate ? new Date(eventData.endDate) : startDate;

      // Create the event - AUTO APPROVE scraped events
      const newEvent = await prisma.event.create({
        data: {
          externalId: eventData.externalId,
          name: eventData.name,
          description: eventData.description || '',
          startDate: startDate,
          endDate: endDate,
          locationName: eventData.locationName || '',
          url: eventData.url,
          cleCredits: eventData.cleCredits,
          status: 'APPROVED', // Auto-approve scraped events
          submittedBy: eventData.submittedBy || 'system@migration',
          submittedAt: new Date(),
          updatedAt: new Date(),
          updatedBy: 'system@migration',
          notes: `Auto-approved scraped event migrated from JSON file on ${new Date().toISOString()}`,
          locationId: locationId,
          communityId: communityId
        }
      });

      console.log(`Created and approved event: ${eventData.name} (ID: ${newEvent.id})`);
      return true;

    } catch (error) {
      console.error(`Error migrating event ${eventData.name || 'Unknown'}:`, error.message);
      return false;
    }
  }

  async migrateAllEvents() {
    console.log('Loading events from JSON file...');
    const events = await this.loadEventsFromJson();

    if (!events.length) {
      console.log('No events found in JSON file!');
      return;
    }

    console.log(`Found ${events.length} events to migrate`);

    let successCount = 0;
    let errorCount = 0;

    for (let i = 0; i < events.length; i++) {
      const event = events[i];
      console.log(`Processing event ${i + 1}/${events.length}: ${event.name || 'Unknown'}`);

      if (await this.migrateEvent(event)) {
        successCount++;
      } else {
        errorCount++;
      }
    }

    console.log('\nMigration complete!');
    console.log(`Successfully migrated and approved: ${successCount} events`);
    console.log(`Errors: ${errorCount} events`);
  }
}

async function main() {
  const migrator = new EventMigrator();

  try {
    console.log('Starting migration...');
    await migrator.migrateAllEvents();
  } catch (error) {
    console.error('Error during migration:', error);
    process.exit(1);
  } finally {
    await prisma.$disconnect();
  }
}

if (require.main === module) {
  main();
} 

File: scripts/migrate-events-to-db.py
#!/usr/bin/env python3
"""
Script to migrate events from JSON files to the database.
This script reads events from public/data/all_events_combined.json and inserts them into the database.
"""

import json
import os
import sys
from datetime import datetime
from typing import Dict, Any, List
import asyncio
from prisma import Prisma

# Add the project root to the path so we can import from src
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

class EventMigrator:
    def __init__(self):
        self.prisma = Prisma()
        self.data_file = os.path.join(os.path.dirname(os.path.dirname(__file__)), "public", "data", "all_events_combined.json")
        
    async def connect(self):
        """Connect to the database."""
        await self.prisma.connect()
        
    async def disconnect(self):
        """Disconnect from the database."""
        await self.prisma.disconnect()
        
    def load_events_from_json(self) -> List[Dict[str, Any]]:
        """Load events from the JSON file."""
        if not os.path.exists(self.data_file):
            print(f"Error: {self.data_file} not found!")
            return []
            
        with open(self.data_file, 'r', encoding='utf-8') as f:
            data = json.load(f)
            return data.get('events', [])
    
    async def create_or_get_community(self, community_name: str) -> str:
        """Create or get a community by name."""
        if not community_name:
            return None
            
        # Try to find existing community
        existing = await self.prisma.community.find_first(
            where={"name": community_name}
        )
        
        if existing:
            return existing.id
            
        # Create new community
        new_community = await self.prisma.community.create({
            "data": {
                "name": community_name,
                "description": f"Community for {community_name}"
            }
        })
        
        return new_community.id
    
    async def create_or_get_location(self, location_data: Dict[str, Any]) -> str:
        """Create or get a location by address."""
        if not location_data or not location_data.get('address'):
            return None
            
        address = location_data['address']
        
        # Try to find existing location
        existing = await self.prisma.location.find_first(
            where={"address": address}
        )
        
        if existing:
            return existing.id
            
        # Create new location
        new_location = await self.prisma.location.create({
            "data": {
                "name": location_data.get('name', 'Unknown Location'),
                "address": address,
                "city": location_data.get('city', ''),
                "state": location_data.get('state', ''),
                "zip": location_data.get('zip', '')
            }
        })
        
        return new_location.id
    
    async def migrate_event(self, event_data: Dict[str, Any]) -> bool:
        """Migrate a single event to the database."""
        try:
            # Check if event already exists by externalId or name + startDate
            existing_event = None
            if event_data.get('externalId'):
                existing_event = await self.prisma.event.find_unique(
                    where={"externalId": event_data['externalId']}
                )
            
            if not existing_event:
                # Try to find by name and start date
                start_date = datetime.fromisoformat(event_data['startDate'].replace('Z', '+00:00'))
                existing_event = await self.prisma.event.find_first(
                    where={
                        "name": event_data['name'],
                        "startDate": start_date
                    }
                )
            
            if existing_event:
                print(f"Event already exists: {event_data['name']}")
                return True
            
            # Create or get community
            community_id = None
            if event_data.get('community'):
                community_id = await self.create_or_get_community(event_data['community'])
            
            # Create or get location
            location_id = None
            if event_data.get('location'):
                location_id = await self.create_or_get_location(event_data['location'])
            
            # Parse dates
            start_date = datetime.fromisoformat(event_data['startDate'].replace('Z', '+00:00'))
            end_date = datetime.fromisoformat(event_data['endDate'].replace('Z', '+00:00')) if event_data.get('endDate') else start_date
            
            # Create the event
            new_event = await self.prisma.event.create({
                "data": {
                    "externalId": event_data.get('externalId'),
                    "name": event_data['name'],
                    "description": event_data.get('description', ''),
                    "startDate": start_date,
                    "endDate": end_date,
                    "locationName": event_data.get('locationName', ''),
                    "url": event_data.get('url'),
                    "cleCredits": event_data.get('cleCredits'),
                    "status": "PENDING",  # Start as pending for admin review
                    "submittedBy": event_data.get('submittedBy', 'system@migration'),
                    "submittedAt": datetime.now(),
                    "updatedAt": datetime.now(),
                    "updatedBy": 'system@migration',
                    "notes": f"Migrated from JSON file on {datetime.now().isoformat()}",
                    "locationId": location_id,
                    "communityId": community_id
                }
            })
            
            print(f"Created event: {event_data['name']} (ID: {new_event.id})")
            return True
            
        except Exception as e:
            print(f"Error migrating event {event_data.get('name', 'Unknown')}: {e}")
            return False
    
    async def migrate_all_events(self):
        """Migrate all events from JSON to database."""
        print("Loading events from JSON file...")
        events = self.load_events_from_json()
        
        if not events:
            print("No events found in JSON file!")
            return
        
        print(f"Found {len(events)} events to migrate")
        
        success_count = 0
        error_count = 0
        
        for i, event in enumerate(events, 1):
            print(f"Processing event {i}/{len(events)}: {event.get('name', 'Unknown')}")
            
            if await self.migrate_event(event):
                success_count += 1
            else:
                error_count += 1
        
        print(f"\nMigration complete!")
        print(f"Successfully migrated: {success_count} events")
        print(f"Errors: {error_count} events")

async def main():
    """Main function."""
    migrator = EventMigrator()
    
    try:
        print("Connecting to database...")
        await migrator.connect()
        
        print("Starting migration...")
        await migrator.migrate_all_events()
        
    except Exception as e:
        print(f"Error during migration: {e}")
        sys.exit(1)
    finally:
        await migrator.disconnect()

if __name__ == "__main__":
    asyncio.run(main()) 